hack/test-go.sh

```diff
@@ -1,124 +1,173 @@
+print_packages="${PRINT_PACKAGES:-}"
+test_kube="${TEST_KUBE:-}"
+verbose="${VERBOSE:-}"
+test_timeout="${TIMEOUT:-60s}"
+detect_races="${DETECT_RACES:-true}"
+coverage_output_dir="${COVERAGE_OUTPUT_DIR:-}"
+coverage_spec="${COVERAGE_SPEC:--cover -covermode atomic}"
+gotest_flags="${GOTEST_FLAGS:-}"
+junit_report="${JUNIT_REPORT:-}"
+
+if [[ -n "${junit_report}" && -n "${coverage_output_dir}" ]]; then
+    echo "$0 cannot create jUnit XML reports and coverage reports at the same time."
+    exit 1
 fi
 
-OUTPUT_COVERAGE=${OUTPUT_COVERAGE:-""}
+# Build arguments for 'go test'
+if [[ -n "${verbose}" || -n "${junit_report}" ]]; then
+    # verbosity can be explicitly set or set implicitly by asking for the jUnit XML report
+    gotest_flags="${gotest_flags} -v"
+fi
 
-if [ -n "${OUTPUT_COVERAGE}" ]; then
-  if [ -z ${KUBE_RACE+x} ]; then
-    KUBE_RACE="-race"
-  fi
-  if [ -z "${KUBE_COVER}" ]; then
-    KUBE_COVER="-cover -covermode=atomic"
-  fi
+if [[ -n "${detect_races}" ]]; then
+    gotest_flags="${gotest_flags} -race"
 fi
 
-if [ -z ${KUBE_RACE+x} ]; then
-  KUBE_RACE=""
+# check to see if user has not disabled coverage mode
+if [[ -n "${coverage_spec}" ]]; then
+    # if we have a coverage spec set, we add it. '-race' assumes '-cover -covermode atomic'
+    # but specifying both at the same time does not lead to an error so we can add both specs
+    gotest_flags="${gotest_flags} ${coverage_spec}"
 fi
 
-KUBE_TIMEOUT=${KUBE_TIMEOUT:-'60s'}
-
-if [ "${1-}" != "" ]; then
-  if [[ "${1}" == *"/..." ]]; then
-    # if the package name ends with /... they are intending a recursive test
-    test_packages=$(find_test_dirs "${1:0:(-4)}")
-  else
-    test_packages="$OS_GO_PACKAGE/$1"
-  fi
-elif [ -n "$TEST_KUBE" ]; then
-  test_packages=`find_test_dirs "."; special_upstream_test_dirs; find_upstream_test_dirs`
-else
-  test_packages=`find_test_dirs "."; special_upstream_test_dirs`
+# check to see if user has not disabled test timeouts
+if [[ -n "${test_timeout}" ]]; then
+    gotest_flags="${gotest_flags} -timeout ${test_timeout}"
 fi
 
-if [ -n "$PRINT_PACKAGES" ]; then
-  for package in $test_packages
-  do
-    echo $package
-  done
+# Determine packages to test
+if [[ "$#" -gt 1 ]]; then
+    test_packages="${@:2}"
+else
+    # If no packages are given to test, we need to generate a list of all packages with unit tests
+    openshift_test_packages="$(find . -not \( \
+        \( \
+              -wholename './Godeps' \
+              -o -wholename './_output' \
+              -o -wholename './_tools' \
+              -o -wholename './.git' \
+              -o -wholename './openshift.local.*' \
+              -o -wholename '*/Godeps/*' \
+              -o -wholename './assets/node_modules' \
+              -o -wholename './test/extended' \
+        \) -prune \
+    \) -name '*_test.go' -print0 | xargs -0n1 dirname | sort -u | xargs -n1 printf "${OS_GO_PACKAGE}/%s\n")"  
```

Comment:
I could refactor `hack/util.sh:731 find_files()` to take a `name` parameter and wrap this into an invocation of that method.